src/models/losses.py

```diff
@@ -4,6 +4,327 @@
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
+        mask = (class_ids == class_ids[i]).type(torch.int)
+        mask[
+            i
+        ] = 0  # This ensures that "correct class" index is not included in the mask.
+        masks.append(mask.reshape(1, -1))  # shape: (1, batch)
+
+        numb_words = cap_len_list[i]
+        query_words = (
+            words_emb[i, :, :numb_words].unsqueeze(0).contiguous()
+        )  # shape: (1, D, L), this picks the caption at ith batch index.
+        query_words = query_words.repeat(
+            batch_size, 1, 1
+        )  # shape: (batch, D, L), this expands the same caption for all batch indices.
+
+        c_i = compute_region_context_vector(
+            local_incept_feat, query_words, gamma1
+        )  # Taken from attnGAN paper. shape: (batch, D, L)
+
+        query_words = query_words.transpose(1, 2).contiguous()  # shape: (batch, L, D)
+        c_i = c_i.transpose(1, 2).contiguous()  # shape: (batch, L, D)
+        query_words = query_words.view(
+            batch_size * numb_words, -1
+        )  # shape: (batch * L, D)
+        c_i = c_i.view(batch_size * numb_words, -1)  # shape: (batch * L, D)
+
+        r_i = compute_relevance(
+            c_i, query_words
+        )  # cosine similarity, or R(c_i, e_i) from attnGAN paper. shape: (batch * L, 1)
+        r_i = r_i.view(batch_size, numb_words)  # shape: (batch, L)
+        r_i = torch.exp(r_i * gamma2)  # shape: (batch, L)
+        r_i = r_i.sum(dim=1, keepdim=True)  # shape: (batch, 1)
+        r_i = torch.log(
+            r_i
+        )  # This is image-text matching score b/w whole image and caption, shape: (batch, 1)
+        match_scores.append(r_i)
+
+    masks = torch.cat(masks, dim=0)  # type: ignore
+    masks = torch.BoolTensor(masks).to(device)  # type: ignore
+    match_scores = torch.cat(match_scores, dim=1)  # type: ignore
+
+    match_scores = gamma3 * match_scores  # This corresponds to P(D|Q) from attnGAN.
+    match_scores.data.masked_fill_(  # type: ignore
+        masks, -float("inf")
+    )  # mask out the scores for mis-matched samples
+
+    match_scores_t = match_scores.transpose(  # type: ignore
+        0, 1
+    )  # This corresponds to P(Q|D) from attnGAN.
+
+    l1_w = cross_entropy(
+        match_scores, match_labels  # type: ignore
+    )  # This corresponds to L1_w from attnGAN.
+    l2_w = cross_entropy(
+        match_scores_t, match_labels
+    )  # This corresponds to L2_w from attnGAN.
+
+    global_incept_feat = global_incept_feat.unsqueeze(0)  # shape: (1, batch, D)
+    sent_emb = sent_emb.unsqueeze(0)  # shape: (1, batch, D)
+
+    incept_feat_norm = torch.norm(
+        global_incept_feat, 2, dim=2, keepdim=True
+    )  # shape: (1, batch, 1)
+    sent_emb_norm = torch.norm(sent_emb, 2, dim=2, keepdim=True)  # shape: (1, batch, 1)
```

Comment:
You attached formula for cosine similarity between two Q and D, it's not the formula for norming at least because norming is for one tensor
![image](https://user-images.githubusercontent.com/54360024/198984296-e9cb4cf7-ae94-482b-913d-fd008356bc24.png)

Reply:
we are specifying dimension = 2 argument along which to apply norm so its equivalent to norm for tensor and not matrix from the docs.. i think the formula follows from the paper, if i misunderstood, could you rephrase your question?

Reply:
I completely missed the idea that you implemented cosine similarity from scratch... My new questions is why? PyTorch already has it, you don't have to calculate norms and apply the dot product, just pass it here in a function

source: [pytorch docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html)